lightning/src/ln/peers/chacha.rs

```diff
@@ -3,21 +3,18 @@ use util::chacha20poly1305rfc::ChaCha20Poly1305RFC;
 
 pub const TAG_SIZE: usize = 16;
 
-pub fn encrypt(key: &[u8], nonce: u64, associated_data: &[u8], plaintext: &[u8]) -> Vec<u8> {
+pub fn encrypt(key: &[u8], nonce: u64, associated_data: &[u8], plaintext: &[u8], ciphertext_out: &mut [u8]) {
 	let mut nonce_bytes = [0; 12];
 	nonce_bytes[4..].copy_from_slice(&byte_utils::le64_to_array(nonce));
 
 	let mut chacha = ChaCha20Poly1305RFC::new(key, &nonce_bytes, associated_data);
-	let mut ciphertext = vec![0u8; plaintext.len()];
 	let mut authentication_tag = [0u8; 16];
-	chacha.encrypt(plaintext, &mut ciphertext, &mut authentication_tag);
+	chacha.encrypt(plaintext, &mut ciphertext_out[..plaintext.len()], &mut authentication_tag);
 
-	let mut tagged_ciphertext = ciphertext;
-	tagged_ciphertext.extend_from_slice(&authentication_tag);
-	tagged_ciphertext
+	ciphertext_out[plaintext.len()..].copy_from_slice(&authentication_tag);
 }
 
-pub fn decrypt(key: &[u8], nonce: u64, associated_data: &[u8], tagged_ciphertext: &[u8]) -> Result<Vec<u8>, String> {
+pub fn decrypt(key: &[u8], nonce: u64, associated_data: &[u8], tagged_ciphertext: &[u8], plaintext_out: &mut [u8]) -> Result<(), String> {
```

Comment:
Same here

Reply:
One of the challenges with performance (in this case memory allocation work) is you have to be surgical with how you use the libraries to get the behavior you want out. In this case, it means using stack-based arrays during the handshake and variable-sized, afterward. I am dubious that modifying input values is unidiomatic. It might not be as pretty, but that is the tradeoff you make when you want to use the same common code for two different use cases.